### PR TITLE
refactor iasmdmd.asm_emit

### PR DESCRIPTION
**Pull request by benjones**
Replace some gotos with functions, eliminate others.  Shrink some variable scopes, remove some duplicated code.  Probably easier to understand commit by commit

### CODE REVIEW

File: src/dmd/iasmdmd.d [hunk 1]
```diff
@@ -1245,33 +1245,115 @@ code *asm_emit(Loc loc,
...
+                    pc.IFL2 = FLconst;
+                }
+                break;
+
+            default:
+                break;
+        }
+    }
+
+    static code* finalizeCode(Loc loc, code* pc, PTRNTAB ptb){
```
Comment (thewilsonator) [inline src/dmd/iasmdmd.d:1317]: brace on newline

File: src/dmd/iasmdmd.d [hunk 2]
```diff
@@ -1253,25 +1253,24 @@ code *asm_emit(Loc loc,
...
-    uint            uSizemaskTable1 =0, uSizemaskTable2 =0,
-                        uSizemaskTable3 =0;
-    ASM_OPERAND_TYPE    aoptyTable1 = _reg, aoptyTable2 = _reg, aoptyTable3 = _reg;
-    ASM_MODIFIERS       amodTable1 = _normal,
-                        amodTable2 = _normal;
-    uint            uRegmaskTable1 = 0, uRegmaskTable2 =0;
+    ASM_MODIFIERS[2] amods = _normal;
+    uint[3] uSizemaskTable = 0;
+    ASM_OPERAND_TYPE[3] aoptyTable = _reg;
+    ASM_MODIFIERS[2] amodTable = _normal;
```
Comment (n8sh) [inline src/dmd/iasmdmd.d:1254]: What is the point of this "array-ification"?

Comment (benjones) [reply src/dmd/iasmdmd.d:1254]: They allow factoring setRegisterProperties.  I hoped I'd find more places where repeated code could turn into loops or functions (I suspect there are more but this PR was getting a bit big anyway).  I also think logically they should be arrays as they are a series of related values

**Review by thewilsonator — APPROVED** [review]

**Comment by WalterBright** [comment]
@thewilsonator @benjones these large refactorings should be done in smaller pieces. Sometimes, there are mistakes in refactorings, and large ones can make it very hard to identify what went wrong.

It's really not a problem to have a sequence of smaller PRs.

**Comment by benjones** [comment]
@WalterBright I tried to make the individual commits stand on their own, but can do multiple PRs in the future

**Comment by WalterBright** [comment]
Thank you. I am regularly reminded the hard way from time to time to keep the PRs small.

**Pull request merged**